File: src/main/java/com/leet/medium/SortList.go
package main

/*
148. Sort List
https://leetcode.com/problems/sort-list/
Medium

Given the head of a linked list, return the list after sorting it in ascending order.
Follow up: Can you sort the linked list in O(n logn) time and O(1) memory (i.e. constant space)?

Example 1:
Input: head = [4,2,1,3]
Output: [1,2,3,4]

Example 2:
Input: head = [-1,5,3,4,0]
Output: [-1,0,3,4,5]

Example 3:
Input: head = []
Output: []

Constraints:
    The number of nodes in the list is in the range [0, 5 * 104].
    -105 <= Node.val <= 105
*/

/**
 * Definition for singly-linked list.
 */
type ListNode struct {
	Val  int
	Next *ListNode
}

func sortList(head *ListNode) *ListNode {
	slow, fast := head, head
	var prev *ListNode
	for fast != nil && fast.Next != nil {
		fast = fast.Next.Next
		prev = slow
		slow = slow.Next
	}

	if fast == slow {
		return head
	}

	otherHead := slow
	if prev != nil {
		prev.Next = nil
	}

	sortedA, sortedB := sortList(head), sortList(otherHead)

	sentinel := &ListNode{}
	node := sentinel

	insAsTail := func(a *ListNode, b *ListNode) (*ListNode, *ListNode) {
		a.Next = b
		b = b.Next
		a = a.Next
		return a, b
	}

	for sortedA != nil && sortedB != nil {
		if sortedA.Val < sortedB.Val {
			node, sortedA = insAsTail(node, sortedA)
		} else {
			node, sortedB = insAsTail(node, sortedB)
		}
	}

	for sortedA != nil {
		node, sortedA = insAsTail(node, sortedA)
	}

	for sortedB != nil {
		node, sortedB = insAsTail(node, sortedB)
	}

	return sentinel.Next
}

File: src/main/java/com/leet/medium/SortList.rs
// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
struct ListNode {
  pub val: i32,
  pub next: Option<Box<ListNode>>,
}

impl ListNode {
  #[inline]
  fn new(val: i32) -> Self {
    ListNode { next: None, val }
  }
}

struct Solution {}

impl Solution {
  pub fn sort_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    let sentinel = ListNode::new(0);

    let mut traverse = &mut head;

    while traverse.is_some() {
      let mut node = &sentinel.next;
      let mut prev = &sentinel;
      let ref_traverse = traverse.as_ref().unwrap();
      while node.is_some() {
        let ref_node = node.as_ref().unwrap();
        if ref_node.val < ref_traverse.val {
          break;
        }
        prev = ref_node;
        node = &ref_node.next;
      }
      let tmp = ref_traverse;
      prev.next = {..*traverse};
      // traverse.as_mut().unwrap().next = node;
      // traverse = &mut tmp.next;
      // prev.next = ref_traverse;
    }
    sentinel.next
  }
}

fn main() {}
